### ISSUE TITLE
Question: Manage multiple progress bars in a multiprocessing environment

### ISSUE THREAD

**Issue by jpfeuffer** [willmcgugan/rich#146]
Hi,

first, thanks a lot for this beautiful library.
I wanted to know if it is possible to manage a Progress object with multiple progress bars in a `multiprocessing` environment.
I just want to have a very basic progress bar for every process.
My first tries did not succeed due to the fact that the Progress object cannot be pickled (due to its usage of an RLock?).
`TypeError: can't pickle _thread.RLock objects`.

**Comment by willmcgugan** [comment]
Hi @jpfeuffer 

Glad you like rich.

I'm afraid that even if you could pickle Progress objects, it wouldn't work. The problem is that even with multiple processes, you only have one stdout stream.

The solution is to have a single Progress instance, defined inside your `if __name__ == __main__` block. Create tasks with `Progress.add_task` and call each process with the returned task ID. You'll then need to post updates from your processes back to the main process, via a "pipe" or "queue".

You should also be aware that you won't be able to print anything from your sub-processes, or it will break the visuals. Anything you want to print should be sent to the main process.

Hope that helps!

**Comment by jpfeuffer** [comment]
Thanks a lot. It did. It worked fine after re-structuring and using a Pipe and letting the spawned processes
send their updates to the main process with their task_id.

**Comment by ChristianMichelsen** [comment]
Hi (and sorry for replying in this quite old thread).



@willmcgugan you wrote the following above:
> You'll then need to post updates from your processes back to the main process, via a "pipe" or "queue".

Do you have any idea or examples on how to do this? 

---

I have a a function something along the line of this:

```python

def worker(filenames):

    # this is a very slow initialisation
    init = very_slow_initialisation()

    # loop over files
    for filename in filenames:

        # quite fast function once the initialisation is done
        fast_function(filename, init)
```

So it is a function that is quite slow to initialise, but once initialised, each loop is takes around a second. To speed it up, I split the filenames into `N` chunks, where `N` is the same as the total number of cores, to maximise the parallelisation, while decreasing number of times the slow initialisation has to run. 

I would like to use a progress bar on the loop, just for a single one of the cores (since all the chunks takes approximately the same time to process). 
In this case I cannot use any of the previous examples with multiprocessing that I have been able to find (e.g. [#121](https://github.com/willmcgugan/rich/issues/121#issuecomment-648716778)), since they update the progress bar each time a new task is sent to the worker process, which only happens once in total in my case.

Do you have any suggestions or tips?

**Comment by willmcgugan** [comment]
Hi @ChristianMichelsen 

It's hard to say without know the full details. But the most efficient way of tackling this would probably be with a queue. Basically create a queue in the main process and `put()` all the filenames in it. Your workers should call `queue.get()` to get and process filenames until the queue is empty. You can update the progress bar from the main process by monitoring the length of the queue, which will eventually get 0 and you can exit your main process.

The alternative is a pipe. When your worker completes a filename, you can call `connection.send('done')`. The main process gets that message via `connection.recv` and advances the progress bar.

**Comment by ChristianMichelsen** [comment]
@willmcgugan Thanks a lot for your answer. After a bit of research, debugging, and generally playing around with the `multiprocessing` module and got the following to work:

```python
def worker(queue_in, queue_out):
    init = very_slow_initialisation()
    while True:
        filename = queue_in.get(block=True)
        if filename is None:
            break
        result = fast_function(filename, init)
        queue_out.put((filename, result))


def main(df):

    queue_in = Queue()
    queue_out = Queue()
    the_pool = Pool(NUM_CORES, worker, (queue_in, queue_out))

    filenames = get_filenames()
    N = len(filenames)

    with Progress() as progress: 
        task_id = progress.add_task("Task", total=N)

        for filename in filenames:
            queue_in.put(filename)

        # Get and print results
        results = []
        for _ in range(N):
            results.append(queue_out.get())
            progress.advance(task_id)

    for _ in range((N):
        queue_in.put(None)

    # prevent adding anything more to the queue and wait for queue to empty
    queue_in.close()
    queue_in.join_thread()

    # prevent adding anything more to the process pool and wait for all processes to finish
    the_pool.close()
    the_pool.join()

    return results
```

**Comment by willmcgugan** [comment]
Good solution!
